src/index.mjs

```diff
@@ -35,85 +35,78 @@ const RETURN = 13;
 const SPACE = 32;
 const SLASH = 47;
 
-const MODE_WHITESPACE = 0;
-const MODE_TEXT = 1;
+const MODE_TEXT = 0;
+const MODE_WHITESPACE = 1;
 const MODE_TAGNAME = 9;
 const MODE_ATTRIBUTE = 13;
 const MODE_SKIP = 47;
 
 /** Create a template function given strings from a tagged template. */
 const build = (statics) => {
-	let mode = MODE_WHITESPACE;
+	let mode = MODE_TEXT;
 	let out = 'return ';
 	let buffer = '';
-	let field = '';
-	let hasChildren = 0;
+	let tagClose = '';
+	let childClose = '';
 	let props = '';
 	let propsClose = '';
 	let spreadClose = '';
 	let quote = 0;
-	let slash, charCode, inTag, propName, propHasValue;
+	let fallbackPropValue = true;
+	let charCode, propName;
 
-	const commit = () => {
-		if (!inTag) {
+	const commit = field => {
+		if (mode === MODE_TEXT) {
 			if (field || (buffer = buffer.replace(/^\s*\n\s*|\s*\n\s*$/g,''))) {
-				if (hasChildren++) out += ',';
-				out += field || stringify(buffer);
+				out += childClose + (field || stringify(buffer));
+				childClose = ',';
 			}
 		}
 		else if (mode === MODE_TAGNAME) {
-			if (hasChildren++) out += ',';
-			out += 'h(' + (field || stringify(buffer));
+			out += childClose + 'h(' + (field || stringify(buffer));
+			childClose = ',';
 			mode = MODE_WHITESPACE;
 		}
-		else if (mode === MODE_ATTRIBUTE || (mode === MODE_WHITESPACE && buffer === '...')) {
-			if (mode === MODE_WHITESPACE) {
-				if (!spreadClose) {
-					spreadClose = ')';
-					if (!props) props = 'Object.assign({}';
-					else props = 'Object.assign(' + props;
-				}
-				props += propsClose + ',' + field;
-				propsClose = '';
+		else if (mode === MODE_WHITESPACE && buffer === '...') {
+			if (!spreadClose) {
+				props = 'Object.assign(' + (props || '{}');
 			}
-			else if (propName) {
-				if (!props) props += '{';
-				else props += ',' + (propsClose ? '' : '{');
+			props += propsClose + ',' + (field || '');
+			propsClose = '';
+			spreadClose = ')';
+		}
+		else if (mode === MODE_ATTRIBUTE || mode === MODE_WHITESPACE) {
+			if (mode === MODE_WHITESPACE) propName = buffer, buffer = '';
```

Comment:
lol Terser does this for us if you want to use two statements :P

Reply:
Also: is this missing the switch from whitespace back to attribute from L90?

Reply:
The old code (in L90) switched to `MODE_ATTRIBUTE`, then called `commit` and then switched back to `MODE_WHITESPACE`. This new version reorganizes the logic so that the "recursive" `commit` call can be skipped.

Reply:
> lol Terser does this for us if you want to use two statements :P

Cool, I'll fix that!